python-lib/pagination.py

```diff
@@ -62,27 +67,49 @@ def set_counting_key(self, counting_key):
         self.counting_key = counting_key
         logger.info("set_counting_key: counting_key set to {}".format(self.counting_key))
 
-    def update_next_page(self, data, response_links=None):
-        response_links = response_links or {}
-        next_link = response_links.get('next', {})
-        next_page_url = next_link.get("url")
+    def update_next_page_offset(self, data, response_links=None):
+        self.is_first_batch = False
+        self.counter += 1
+        self.next_page_number = self.next_page_number + 1
+        self.data_is_list = False
+        if isinstance(data, list):
+            self.data_is_list = True
+            batch_size = len(data)
+            self.records_to_skip = self.records_to_skip + batch_size
+            if batch_size == 0:
+                self.is_last_batch_empty = True
+            logger.info("update_next_page_offset:data is list:batch_size={}, records_to_skip={}, is_last_batch_empty={}".format(
+                batch_size, self.records_to_skip, self.is_last_batch_empty
+            ))
+            return
+        elif self.counting_key:
+            extracted_data = get_value_from_path(data, self.counting_key.split("."), can_raise=False)
+            if extracted_data:
+                batch_size = len(extracted_data)
+            else:
+                batch_size = 0
+                self.is_last_batch_empty = True
+        else:
+            batch_size = 1
+        if self.skip_key:
+            self.skip = data.get(self.skip_key)
+            logger.info("update_next_page_offset:skip=data[{}]={}".format(self.skip_key, self.skip))
+        self.records_to_skip = self.records_to_skip + batch_size
+        logger.info("update_next_page_offset:records_to_skip={}, batch_size={}".format(self.records_to_skip, batch_size))
+
+
+    def update_next_page_per_page(self, data, response_links=None):
```

Comment:
I think it *might* be safe/wise to factor out a method to calculate batch_size from both the per page and offset methods. That's make them easier to read too. But would need to be careful.

Reply:
Done in [3b46290](https://github.com/dataiku/dss-plugin-api-connect/pull/30/commits/3b46290e47409bd30b45e5a541d452a6fff32bde)